### PR TITLE
Autoscaling

### PR DESCRIPTION
**Pull request by joshski**
Support autoscaling group deployments. This sucks, but they need a separate CodeDeploy deployment group, which means we need to know for any given ApplicationSet subdirectory, which ones target auto scaling groups. So we now have a mandatory `deployspec.yml`...

Lots of refactoring, bug fixes (while loops did not all have sleeps in them) and changed the way we aggregate feedback from the set of deployments (previous version would fail on the first failed deployment, not wait for all of them).

Sanity check please @PaulCampbell / @Jetski5822

### CODE REVIEW

**Comment by joshski** [comment]
Needs some documentation, doing that now...

File: TTC.Deployment.AmazonWebServices/Deployer.cs [hunk 1]
```diff
@@ -160,114 +149,19 @@ void EnsureCodeDeployRoleExists()
...
-            return deploymentResponse;
-        }
-
         void WaitForBundlesToDeploy(List<string> deploymentIds)
         {
             var  deploymentsInfo = new List<DeploymentInfo>() ;
             var inProgress = deploymentIds.Count;
             while (inProgress > 0)
             {
+                Thread.Sleep(TimeSpan.FromSeconds(10));
```
Comment (PaulCampbell) [inline TTC.Deployment.AmazonWebServices/Deployer.cs]: ha. good spot

File: TTC.Deployment.Tests/AutoScalingDeploymentTest.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                StackName = StackName,
+                TemplatePath = @".\example-windows-vpc-autoscaling-group-template.json"
+            });
+            _hasCreatedStack = true;
+        }
+
+        [TestFixtureTearDown]
+        public void TearDown()
+        {
+            // DeletePreviousTestStack();
```
Comment (PaulCampbell) [inline TTC.Deployment.Tests/AutoScalingDeploymentTest.cs]: Maybe drop this back in?

File: TTC.Deployment.Tests/DeploymentTest.cs [hunk 3]
```diff
@@ -45,9 +45,9 @@ public void EnsureStackExists()
         }
 
         [TestFixtureTearDown]
-        public void TearDown()
+        public void TestFixtureTearDown()
         {
-            DeletePreviousTestStack();
+            // DeletePreviousTestStack();
```
Comment (PaulCampbell) [inline TTC.Deployment.Tests/DeploymentTest.cs]: same

**Comment by joshski** [comment]
Fixed as per @PaulCampbell's comments... good to merge?

@marko-baric-travcorp any comments?

**Comment by marko-baric-travcorp** [comment]
@joshski Looks good to me

**Comment by PaulCampbell** [comment]
2 secs - just trying to VS installed.  Basically interested in how this thing behaves with a missing deployspec.yml file (that thing, whilst necessary is a bit strange so a friendly exception if it's missing is what I'm looking for - any thoughts?)

**Comment by joshski** [comment]
Yeah I'm pretty sure it says something along the lines of "couldn't find file at <fullpath>/deployspec.yml" but it could throw a specific exception...

**Pull request merged**